### PR TITLE
units: Restore the +1 in relative locktime satisfied by height

### PR DESCRIPTION
**Pull request by kwsantiago**
#6582 removed the `+ 1` from the height based relative locktime check. It should be restored.

Absolute and relative locktimes differ in consensus. `CheckFinalTxAtTip` compares `nLockTime` against the next block height, so the absolute half of #6582 is correct. But `CalculateSequenceLocks` computes

    nMinHeight = nCoinHeight + relative - 1

and `EvaluateSequenceLocks` requires `nMinHeight < block.nHeight`. For the next block that reduces to `relative <= diff + 1`, so the UTXO needs `relative` confirmations, not `relative + 1`.

On master every height based relative locktime reads as unsatisfied for one block longer than consensus requires.

Checked against Bitcoin Core v31.1 on regtest, sending version 2 transactions to `testmempoolaccept` at each tip:

    N   utxo height   first tip Core accepts   confirmations
    1   151           151                      1
    2   156           157                      2
    5   162           166                      5
    9   171           179                      9

Core accepts at N confirmations, master requires N + 1. Over 29 (N, tip) pairs there are 4 mismatches on master and 0 with this change. Same result on v28.0. The same check run against the absolute code, untouched here, gives 0 mismatches on master.

The time based path is left alone. `EvaluateSequenceLocks` compares `nMinTime` against the median time past of the chain tip, not of the next block, so it correctly has no `+ 1` (24 pairs with bit 22 set, 0 mismatches on master). A comment records the asymmetry.

Also reverts the three test constants changed by #6582 and adds a boundary test.

### CODE REVIEW

File: units/src/locktime/relative/mod.rs [hunk 1]
```diff
@@ -440,9 +440,20 @@ impl NumberOfBlocks {
...
+            // it to be _less than_ the height of the block containing the spend, which is
+            // `chain_tip + 1`. That reduces to `self <= diff + 1`, i.e. the UTXO must have at least
+            // `self` confirmations once the next block is mined.
+            //
+            // Note this is deliberately asymmetric with the time based check below, which compares
+            // against the MTP of `chain_tip` itself rather than of the next block and so has no
+            // `+ 1`. Do not "unify" the two.
+            //
+            // Use `u64` so that `diff + 1` cannot overflow when `diff` is `u32::MAX`.
+            .map(|diff| u64::from(u32::from(self.to_height())) <= u64::from(diff.to_u32()) + 1)
```
Comment (tcharding) [inline units/src/locktime/relative/mod.rs]: We don't need `u64`. Height was a `u16` so +1 cannot overflow a `u32`.

Also remove all the code comments.

**Comment by tcharding** [comment]
Wow this is cool. We merged yesterday and this bot found this bug already. Someone is white hat scanning our repo hard.

FWIW I spent about half an hour reviewing this to convince myself it was indeed correct. I accept full responsibility for introducing the bug. I forgot about this difference between relative and absolute lock times when doing #6582. I maintain my position that the locktime code is _very_ difficult to patch and review. We should continue to view all changes to the module with suspicion, even when done by one of the maintainers.

**Comment by apoelstra** [comment]
~~Jeez, I can't believe I let this through. Your commit comment in #6582 says "either the +1 or <= is correct, but not both". You initially removed the <= and replaced the +1 with a `saturating_add`, which was wrong and I complained about. Then you removed the `saturating_add` but *didn't put the +1 back*.~~ Never mind.

**Comment by apoelstra** [comment]
@kwsantiago can you change your test to also include `n = 0`?

It's surprising that we don't have an existing test for this. Obviously if the relative locktime is zero then you're allowed to put a transaction and its child into the same block. Our broken logic would've forbidden this, which we should have been testing (but clearly weren't).

**Comment by apoelstra** [comment]
Specifically change the line `for n in [1_u16, 2, 5, 9, 20]` to `for n in [0, 2, 5, 9, 20, u16::MAX]`. We should hit 0 and also hit the maximum, since we've repeatedly gone back and forth about saturating arithmetic.

Your loop body even has a `if n > 1` block but right now you never set `n = 0`.

**Comment by kwsantiago** [comment]
Both applied, thanks.

One correction: `n = 0` was never broken. The old check was `0 <= diff`, which holds for every `u32`, so zero always passed and the failure only starts at one. The missing coverage was real, the bug wasn't.

Verified on regtest against Core v31.1. `nSequence=0` is accepted at `diff = 0`, and an unconfirmed parent plus a child spending it are accepted together as a package. At the other end, `nSequence=65535` is rejected as `non-BIP68-final` at tip 65725 and accepted at 65726, i.e. exactly 65535 confirmations. 37 (n, tip) pairs total, no mismatches; master misses 6 of them including the max.

On `u64`: the `+ 1` was on `diff`, not the height, so it overflows at `u32::MAX` and fails the existing `test_max_height_satisfaction`. Moving the `- 1` to the height side avoids both that and the `u64`.

Separately, the same-block case can't be expressed through this API at all. It needs `utxo_mined_at = chain_tip + 1`, which returns `InvalidHeightError` with a message suggesting the arguments are the wrong way around. Happy to open a separate issue.

On the bot comment: I did use AI to help find this, but that doesn't make me a bot. I checked all of it against a regtest node rather than taking the tool's word for it.

**Comment by apoelstra** [comment]
Thanks for reacting so quickly!

So, the corrected code uses `saturating_sub` which is a nop when `n = 0`, so it makes sense that that case was not broken before. But the new code treats `n = 0` and `n = 1` identically, which feels wrong to me. 0 should mean "can be included in the same block as its parent" while 1 should mean "must appear in a later block than its parent".

I think that for a single transaction whose parent is confirmed `testmempoolaccept` will behave identically for 0 and 1, but for a transaction whose parent is still in the mempool, they should behave differently.

Now, this API takes `utxo_mined_at` so I guess we can't actually express "parent is still in the mempool"...I wonder if we should allow the "chain tip" to actually be one higher than the current block? Or if we should have a different method? Frustrating. But I think this PR, as far as it goes, is correct.

**Comment by kwsantiago** [comment]
They do differ on regtest, but only for an unconfirmed parent:

```
nSequence=0: child accepted
nSequence=1: child rejected, non-BIP68-final
```

With a confirmed parent they are indistinguishable, since Core allows 0 from block h and 1 from h+1, and h is already mined. That matches the API here: 0 and 1 agree on every representable input, and the only one where they would differ is `chain_tip = utxo_mined_at - 1`, which `checked_sub` rejects.

**Comment by apoelstra** [comment]
Cool. So I think the existing API is okay (with your fix).

@tcharding I'm a liitle tempted to relax the "check that `mined_at` and `chain_tip` are in the right order" logic to allow `mined_at` to be one past the chain tip. Users could use this to indicate an unconfirmed parent. What do you think?)

Regardless we should extend the docs to talk about unconfirmed transactions.

**Review by tcharding — APPROVED** [review]
ACK ee13c6bb72b302721f721167c3c7d8fbca416a44

**Review by apoelstra — APPROVED** [review]
ACK ee13c6bb72b302721f721167c3c7d8fbca416a44; successfully ran local tests
